### PR TITLE
Add cluster member evacuation

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

File: lxc/cluster.go [hunk 1]
```diff
@@ -913,3 +922,117 @@ func (c *cmdClusterUpdateCertificate) Run(cmd *cobra.Command, args []string) err
...
+		return errors.Wrap(err, "Failed to parse servers")
+	}
+
+	resource := resources[0]
+
+	if resource.name == "" {
+		return fmt.Errorf(i18n.G("Missing cluster member name"))
+	}
+
+	if !c.flagForce && !cli.AskBool(fmt.Sprintf("Are you sure you want to %s cluster member %q? (yes/no) [default=no]: ", cmd.Name(), resource.name), "no") {
```
Comment (stgraber) [inline lxc/cluster.go:998]: The string should be fed to `i18n.G` for translation

File: lxd/instances_post.go [hunk 2]
```diff
@@ -11,7 +11,7 @@ import (
 	"os"
 	"strings"
 
-	"github.com/dustinkirkland/golang-petname"
+	petname "github.com/dustinkirkland/golang-petname"
```
Comment (stgraber) [inline lxd/instances_post.go:14]: Un-needed.

File: lxd/instance_backup.go [hunk 3]
```diff
@@ -134,6 +134,10 @@ func instanceBackupsGet(d *Daemon, r *http.Request) response.Response {
 		return resp
 	}
 
+	if d.cluster.LocalNodeIsEvacuated() {
```
Comment (stgraber) [inline lxd/instance_backup.go:137]: In general I don't think we need all of those changes. If there are instances left on an evacuated server, I don't mind having the ability to still interact with them in theory.

What we need to prevent however is the creation of any new instances, so that part should stay.

**Comment by stgraber** [comment]
Looking good, just left a few comments.
We also need:
 - `make update-api` for a `doc/rest-api: Refresh swagger YAML` commit
 - Update `doc/instances.md` to cover the new config keys

File: lxd/api_cluster.go [hunk 4]
```diff
@@ -2233,3 +2241,531 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+// parameters:
+//   - in: body
+//     name: cluster
+//     description: Cluster member state
+//     required: true
+//     schema:
+//       $ref: "#/definitions/ClusterMemberStatePost"
+// responses:
+//   "200":
+//     $ref: "#/responses/EmptySyncResponse"
```
Comment (stgraber) [inline lxd/api_cluster.go:2265]: This is incorrect, this function returns an Operation, not a SyncResponse.

File: lxd/api_cluster.go [hunk 5]
```diff
@@ -2233,3 +2241,531 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+//   "400":
+//     $ref: "#/responses/BadRequest"
+//   "403":
+//     $ref: "#/responses/Forbidden"
+//   "500":
+//     $ref: "#/responses/InternalServerError"
+func clusterNodeStatePost(d *Daemon, r *http.Request) response.Response {
```
Comment (stgraber) [inline lxd/api_cluster.go:2272]: I would have expected to see some logic here to check `mux.Vars(r)["name"]` against the local name and if it doesn't match, forward the request to the correct server.

As it stands, I'm not seeing forwarding logic, so I suspect that if I hit `/1.0/cluster/members/vm03` on `vm01`, I'll end up with `vm01` getting evacuated rather than `vm03`.

File: lxd/api_cluster.go [hunk 6]
```diff
@@ -2233,3 +2241,531 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+		dbInstances, err = tx.GetInstances(db.InstanceFilter{Node: sourceName})
+		if err != nil {
+			return errors.Wrap(err, "Failed to get instances")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
```
Comment (stgraber) [inline lxd/api_cluster.go:2349]: We need to do that query for every instance, otherwise all instances will be moved to the same target instead of getting properly spread on the remaining servers, potentially causing that one server to go out of resources and crash.

File: lxd/api_cluster.go [hunk 7]
```diff
@@ -2233,3 +2241,531 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+		targetNode, err = tx.GetNodeByName(targetNodeName)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
```
Comment (stgraber) [inline lxd/api_cluster.go:2380]: We need to do that query for every instance, otherwise all instances will be moved to the same target instead of getting properly spread on the remaining servers, potentially causing that one server to go out of resources and crash.

File: lxd/api_cluster.go [hunk 8]
```diff
@@ -2233,3 +2241,531 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+
+	op, err := operations.OperationCreate(d.State(), "", operations.OperationClassTask, db.OperationClusterMemberRestore, nil, nil, run, nil, nil, r)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	return operations.OperationResponse(op)
+}
+
+func getMigratableInstances(d *Daemon, instances []instance.Instance) []instance.Instance {
```
Comment (stgraber) [inline lxd/api_cluster.go:2709]: Feels like this function would better fit within the `instance` package.

File: lxd/api_cluster.go (hunk 5 above)
Comment (monstermunchkin) [reply lxd/api_cluster.go:2272]: The correct cluster member is evacuated or restored in the functions `evacuateClusterMember` and `restoreClusterMember` respectively, without having to forward the request. The test confirms this as we hit `/1.0/cluster/members/node1` on `node2`.

Comment (stgraber) [reply lxd/api_cluster.go:2272]: Ah yeah, I see, you're using the API to shutdown the remote instances. This works but is going to be really slow.

It'd be far preferable to just forward the state change request directly to the correct server which can then load the instances and handle the shutdown operation without any overhead, only establishing remote connections for the migration stage.

File: lxd/db/node.go [hunk 9]
```diff
@@ -690,6 +690,25 @@ func (c *ClusterTx) UpdateNodeFailureDomain(id int64, domain string) error {
...
+		return err
+	}
+
+	n, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	if n != 1 {
+		return fmt.Errorf("query updated %d rows instead of 1", n)
```
Comment (stgraber) [inline lxd/db/node.go:715]: We usually start errors with a capital letter these days.

File: doc/instances.md [hunk 10]
```diff
@@ -41,6 +41,7 @@ boot.autostart.delay                        | integer   | 0                 | n/
 boot.autostart.priority                     | integer   | 0                 | n/a           | -                         | What order to start the instances in (starting with highest)
 boot.host\_shutdown\_timeout                | integer   | 30                | yes           | -                         | Seconds to wait for instance to shutdown before it is force stopped
 boot.stop.priority                          | integer   | 0                 | n/a           | -                         | What order to shutdown the instances (starting with highest)
+cluster.evacuate                            | string    | auto              | n/a           | -                         | What to do when evacuating the instance (auto, migrate, or stop)
```
Comment (stgraber) [inline doc/instances.md:44]: Can you move the doc change to its own commit?

File: lxd/instance/instance_utils.go [hunk 11]
```diff
@@ -13,7 +13,7 @@ import (
 	"github.com/flosch/pongo2"
 	"github.com/pkg/errors"
 
-	"github.com/lxc/lxd/client"
+	lxd "github.com/lxc/lxd/client"
```
Comment (stgraber) [inline lxd/instance/instance_utils.go:16]: Un-needed change

File: lxd/instances_post.go [hunk 12]
```diff
@@ -394,6 +402,10 @@ func createFromMigration(d *Daemon, r *http.Request, projectName string, req *ap
 }
 
 func createFromCopy(d *Daemon, r *http.Request, projectName string, req *api.InstancesPost) response.Response {
+	if d.cluster.LocalNodeIsEvacuated() {
```
Comment (stgraber) [inline lxd/instances_post.go:405]: Don't we have a 4th one of those for `createFromMigration`?

File: lxd/api_cluster.go [hunk 13]
```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+		return response.SmartError(err)
+	}
+
+	// Forward request
+	if localNode != name {
+		resp := forwardedResponseToNode(d, r, name)
+		if resp != nil {
+			return resp
+		}
+	}
```
Comment (stgraber) [inline lxd/api_cluster.go:2296]: Based on what I read about forwardedResponseToNode and ResolveTarget, I don't think you need the if statement.
If the target node is the local node, it will return nil, if it's not, it will forward and return a response.

File: lxd/api_cluster.go [hunk 14]
```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+
+			dest, err := cluster.Connect(targetNode.Address, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+			if err != nil {
+				return errors.Wrap(err, "Failed to connect to destination")
+			}
```
Comment (stgraber) [inline lxd/api_cluster.go:2443]: I don't think you actually ever need to connect to the destination directly.

`source.UpdateInstanceState` should work just fine, LXD will just forward the request internally as needed.

File: lxd/api_cluster.go [hunk 15]
```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+
+	if node.State == db.ClusterMemberStateCreated {
+		return response.SmartError(fmt.Errorf("Node is not evacuated"))
+	}
+
+	// The instances are retrieved in a separate transaction, after the node is in EVACUATED state.
+	var dbInstances []db.Instance
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// If evacuating, consider only the instances on the node which needs to be evacuated.
```
Comment (stgraber) [inline lxd/api_cluster.go:2520]: Bad copy/paste?

File: lxd/api_cluster.go [hunk 16]
```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+		dbInstances, err = tx.GetInstances(db.InstanceFilter{})
```
Comment (stgraber) [inline lxd/api_cluster.go:2521]: Should be `tx.GetInstances(*InstanceFilterAllInstances())`

File: lxd/api_cluster.go [hunk 17]
```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+
+	if node.State == db.ClusterMemberStatePending {
+		return response.SmartError(fmt.Errorf("Cannot restore or restore a pending node"))
+	}
+
+	if node.State == db.ClusterMemberStateCreated {
+		return response.SmartError(fmt.Errorf("Node is not evacuated"))
+	}
+
+	// The instances are retrieved in a separate transaction, after the node is in EVACUATED state.
```
Comment (stgraber) [inline lxd/api_cluster.go:2516]: Wrong comment?

File: lxd/api_cluster.go [hunk 18]
```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+			apiInst, _, err := source.GetInstance(inst.Name())
+			if err != nil {
+				return errors.Wrapf(err, "Failed to get instance %q", inst.Name())
+			}
+
+			if apiInst.Status == "Running" {
+				metadata["evacuation_progress"] = fmt.Sprintf("Stopping %s", inst.Name())
+				op.UpdateMetadata(metadata)
+
+				stopOp, err := source.UpdateInstanceState(inst.Name(), api.InstanceStatePut{Action: "stop", Force: true}, "")
```
Comment (stgraber) [inline lxd/api_cluster.go:2600]: We need to be nice, this would cause dataloss!

File: lxd/api_cluster.go [hunk 19]
```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+	for _, dbInst := range dbInstances {
+		if dbInst.Node == node.Name {
+			inst, err := instance.LoadByProjectAndName(d.State(), dbInst.Project, dbInst.Name)
+			if err != nil {
+				return response.SmartError(errors.Wrap(err, "Failed to load instance"))
+			}
+
+			localInstances = append(localInstances, inst)
+			continue
+		}
```
Comment (stgraber) [inline lxd/api_cluster.go:2544]: This is assuming that all instances should be running, this may not be accurate.

When shutting them down, we should override their last_state.power property to `RUNNING` and then only start back those that were running.

File: lxd/api_cluster.go [hunk 20]
```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+
+		// Only consider instances where volatile.evacuate.origin is set to the node which needs to be restored.
+		val, ok := dbInst.Config["volatile.evacuate.origin"]
+		if !ok || val != node.Name {
+			continue
+		}
```
Comment (stgraber) [inline lxd/api_cluster.go:2550]: Same thing with the remote ones, if we move an instance which is currently stopped, we shouldn't start it.

File: lxd/api_cluster.go [hunk 21]
```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+				continue
+			}
+
+			metadata["evacuation_progress"] = fmt.Sprintf("Stopping %s", inst.Name())
+			op.UpdateMetadata(metadata)
+
+			err = inst.Stop(false)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to stop instance %q", inst.Name())
+			}
```
Comment (stgraber) [inline lxd/api_cluster.go:2413]: We need to be much nicer than this :)

We need logic similar to what we do on clean LXD shutdown, so attempt a normal shutdown of the instance with a timeout, then force it if needed.

**Comment by stgraber** [comment]
Looking pretty good overall. The few remaining points mostly are:
 - Limit number of connections needed to other cluster members
 - Be nicer about shutting down the instances (should do a clean shutdown, timeout then force, respecting `boot.host_shutdown_timeout` if set, otherwise waiting up to 30s)
 - Make sure we maintain instance state. Something that's stopped when being evacuated must remain stopped on its new target and then be restored as stopped. Similarly something evacuated which was running, will keep running on its new target. If it then gets stopped on the new target before it gets restored on the original server, it needs to remain stopped when it arrives back. For that we should just make use of `last_state.power` and set it to RUNNING for instances which were stopped during evacuation.

File: lxd/instances_post.go (hunk 12 above)
Comment (monstermunchkin) [reply lxd/instances_post.go:405]: We cannot restrict `createFromMigration` as we need to function for the migration. If I add the restriction, the test fails.

File: lxd/api_cluster.go (hunk 14 above)
Comment (monstermunchkin) [reply lxd/api_cluster.go:2443]: Tried that but it hangs forever at `startOp.Wait()` because it's waiting for `chActive`. I didn't investigate any further.

Comment (stgraber) [reply lxd/api_cluster.go]: Tried passing `false` instead of `true`?

`true` is for notifications within a cluster and so doesn't get you events, in this case, you're doing a real remote operation so you don't want notify mode.

Comment (monstermunchkin) [reply lxd/api_cluster.go]: If it's set to `false`, I get `Error: Failed to connect to source: Missing event connection with target cluster member`.

Comment (stgraber) [reply lxd/api_cluster.go]: Hmm, well, that error is a bit of a problem as that also explains the event issue.

Are you getting this because the test is running too early and not all members have established event connections yet?

Comment (monstermunchkin) [reply lxd/api_cluster.go]: I see there issue here. Basically, the node is connecting to itself (`cluster.Connect(node.Address, ...)` in order to use `MigrateInstance()`. However, there is not listener with the node's own address which causes the failure.

Comment (stgraber) [reply lxd/api_cluster.go]: Ah, that'd make sense, the whole connecting to itself part is a bit dodgy anyway, it should be able to just make an internal call instead.

File: lxd/api_cluster.go [hunk 22]
```diff
@@ -2233,3 +2242,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+	}
+
+	var targetNodeName string
+	var targetNode db.NodeInfo
+
+	run := func(op *operations.Operation) error {
+		source, err := cluster.Connect(node.Address, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+		if err != nil {
+			return errors.Wrap(err, "Failed to connect to source")
+		}
```
Comment (stgraber) [inline lxd/api_cluster.go:2379]: So instead of connecting to itself over the network, this probably should just call `instancePostClusteringMigrateWithCeph` or `instancePostClusteringMigrate` as appropriate.

I'd recommend moving the chunk of the code deciding between those two functions into its own function, then call that from both `instance_post.go` and from here.

File: lxd/instance/instance_utils.go [hunk 23]
```diff
@@ -1175,3 +1175,68 @@ func NextSnapshotName(s *state.State, inst Instance, defaultPattern string) (str
...
+
+		if val == "stop" {
+			continue
+		}
+
+		devices := inst.ExpandedDevices()
+
+		canMigrate := true
+
+		for _, dev := range devices {
```
Comment (tomponline) [inline lxd/instance/instance_utils.go:1204]: @stgraber @monstermunchkin this feels like we are starting to leak device specific config back into the instances package.

Shouldn't this be a property of the device?

Comment (stgraber) [reply lxd/instance/instance_utils.go:1204]: Yeah, a Migratable() function would be nice to do this cleanly.

Comment (tomponline) [reply lxd/instance/instance_utils.go:1204]: > Yeah, a Migratable() function would be nice to do this cleanly.

Yeah, although I was thinking more an `Inf0()` function like the `Instance` interface, that returns a struct that we could later extend with additional properties if needed. But for now could just have a `Migratable` bool field.

**Comment by monstermunchkin** [comment]
jenkins: test this please

File: shared/instance.go [hunk 24]
```diff
@@ -82,6 +82,10 @@ var InstanceConfigKeysAny = map[string]func(value string) error{
 	"boot.stop.priority":         validate.Optional(validate.IsInt64),
 	"boot.host_shutdown_timeout": validate.Optional(validate.IsInt64),
 
+	"cluster.evacuate": validate.Optional(func(value string) error {
+		return validate.IsOneOf(value, []string{"auto", "migrate", "stop"})
+	}),
```
Comment (stgraber) [inline shared/instance.go:87]: Can't that be made into just `"cluster.evacuate": validate.Optional(validate.IsOneOf(value, []string{"auto", "migrate", "stop"})`?

File: lxd/instances_post.go (hunk 12 above)
Comment (stgraber) [reply lxd/instances_post.go:405]: We can tell whether the request is coming from the cluster or not though.

`r.Context().Value(request.CtxProtocol)` will be set to `cluster` when dealing with an internal request.

File: lxd/instance/instance_utils.go (hunk 23 above)
Comment (stgraber) [reply lxd/instance/instance_utils.go]: Ah yeah, that'd work.

@monstermunchkin can you add an `Info()` function to the `Device` interface and have that return a `DeviceInfo` struct which has a `Migratable` bool? Then update the logic in `driver_common.go` to rely on that rather than the hardcoded device types and details to decide whether an instance is migratable.

File: lxd/instance_post.go [hunk 25]
```diff
@@ -716,3 +686,31 @@ func instancePostCreateContainerMountPoint(d *Daemon, project, containerName str
 
 	return nil
 }
+
+func migrateInstance(d *Daemon, r *http.Request, inst instance.Instance, projectName string, targetNode string, sourceNodeOffline bool, name string, instanceType instancetype.Type, req api.InstancePost) (func(op *operations.Operation) error, error) {
```
Comment (stgraber) [inline lxd/instance_post.go:690]: This looks like a pretty odd signature. Any reason why we can't make `migrateInstance` just:

```
 func migrateInstance(d *Daemon, r *http.Request, inst instance.Instance, projectName string, targetNode string, sourceNodeOffline bool, name string, instanceType instancetype.Type, req api.InstancePost, op *operations.Operation) error {
```

Then you can wrap that into a Run function when called from `instancePost` and can also call it directly from the evacuation code.

File: lxd/api_cluster.go [hunk 26]
```diff
@@ -2233,3 +2242,436 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+		instances[i] = inst
+	}
+
+	var targetNodeName string
+	var targetNode db.NodeInfo
+
+	run := func(op *operations.Operation) error {
+		metadata := make(map[string]interface{})
+
+		// Stop all instances if needed
```
Comment (stgraber) [inline lxd/api_cluster.go:2378]: So one thing we should do differently here. Instead of stopping everything, then moving things and finally starting them back up. We should really go instance by instance and for each do the stop+migrate+start.

That way we minimize downtime by not having to wait for all other instances to be stopped first.

File: lxd/api_cluster.go [hunk 27]
```diff
@@ -2233,3 +2242,436 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+	}
+
+	run := func(op *operations.Operation) error {
+		var source lxd.InstanceServer
+		var sourceNode db.NodeInfo
+
+		metadata := make(map[string]interface{})
+
+		// Migrate instances
+		for _, inst := range instances {
```
Comment (stgraber) [inline lxd/api_cluster.go:2561]: Same comment as during evacuation. When restoring instances, we should do it instance by instance so that each instance is offline for the least amount of time possible.

File: shared/instance.go (hunk 24 above)
Comment (monstermunchkin) [reply shared/instance.go:87]: Nope, function signatures don't match.

File: lxd/instances_post.go (hunk 12 above)
Comment (tomponline) [reply lxd/instances_post.go:405]: @stgraber oh that's new isn't it? I didn't know about that. I thought we still needed to inspect the custom HTTP headers to detect an internal request.

File: shared/instance.go (hunk 24 above)
Comment (stgraber) [reply shared/instance.go:87]: Oh right, I'll send a PR to make it work :)

**Comment by stgraber** [comment]
@monstermunchkin ready for a rebase now

File: lxd/api_cluster.go [hunk 28]
```diff
@@ -2233,3 +2242,434 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+	run := func(op *operations.Operation) error {
+		metadata := make(map[string]interface{})
+
+		// Stop all instances if needed
+		for _, inst := range instances {
+			if inst.IsRunning() {
+				metadata["evacuation_progress"] = fmt.Sprintf("Stopping %s", inst.Name())
+				op.UpdateMetadata(metadata)
+
+				// Be nice
```
Comment (stgraber) [inline lxd/api_cluster.go:2384]: Would change to `Wait up to the host shutdown timeout before forcing instance stop`

File: lxd/api_cluster.go [hunk 29]
```diff
@@ -2233,3 +2242,434 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+				timeout := inst.ExpandedConfig()["boot.host_shutdown_timeout"]
+				val, err := strconv.Atoi(timeout)
+				if err != nil {
+					val = 30
+				}
+
+				err = inst.Shutdown(time.Duration(val) * time.Second)
+				if err != nil {
+					return errors.Wrapf(err, "Failed to stop instance %q", inst.Name())
```
Comment (stgraber) [inline lxd/api_cluster.go:2393]: This should instead attempt a `inst.Stop` at that point. If `inst.Stop` fails, then check if the instance is still running before reporting an error, if it's stopped, just continue.

File: lxd/api_cluster.go [hunk 30]
```diff
@@ -2233,3 +2242,434 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+			if !inst.IsMigratable() {
+				continue
+			}
+
+			metadata["evacuation_progress"] = fmt.Sprintf("Migrating %s", inst.Name())
+			op.UpdateMetadata(metadata)
+
+			inst.VolatileSet(map[string]string{"volatile.evacuate.origin": nodeName})
+
+			startInstance := inst.LocalConfig()["volatile.last_state.power"] == "RUNNING"
```
Comment (stgraber) [inline lxd/api_cluster.go:2408]: You could do a `isRunning := inst.IsRunning()` at the beginning of the loop, then use that both for the if statement and then later to know whether to start the instance back up.

File: lxd/api_cluster.go [hunk 31]
```diff
@@ -2233,3 +2242,434 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+			if err != nil {
+				return errors.Wrap(err, "Failed to get node")
+			}
+
+			source, err = cluster.Connect(sourceNode.Address, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+			if err != nil {
+				return errors.Wrap(err, "Failed to connect to source")
+			}
+
+			source = source.UseTarget("")
```
Comment (stgraber) [inline lxd/api_cluster.go:2568]: This shouldn't be needed

File: lxd/api_cluster.go [hunk 32]
```diff
@@ -2233,3 +2242,434 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+			source = source.UseProject(inst.Project())
+
+			apiInst, _, err := source.GetInstance(inst.Name())
+			if err != nil {
+				return errors.Wrapf(err, "Failed to get instance %q", inst.Name())
+			}
+
+			if apiInst.Status == "Running" {
```
Comment (stgraber) [inline lxd/api_cluster.go:2576]: `apiInst.StatusCode == api.Running` would be cleaner

File: lxd/api_cluster.go [hunk 33]
```diff
@@ -2233,3 +2242,434 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
...
+				metadata["evacuation_progress"] = fmt.Sprintf("Stopping %s", inst.Name())
+				op.UpdateMetadata(metadata)
+
+				stopOp, err := source.UpdateInstanceState(inst.Name(), api.InstanceStatePut{Action: "stop", Force: false}, "")
```
Comment (stgraber) [inline lxd/api_cluster.go:2580]: This should pass the host shutdown timeout as a Timeout or it could be stuck waiting forever.
Same logic as above, if the clean shutdown fails, try a force one, if the force one fails, check if the instance isn't stopped.

File: lxd/instances_post.go (hunk 12 above)
Comment (stgraber) [reply lxd/instances_post.go:410]: Yeah, I think that was added when we did the plumbing for the requestor in the lifecycle events.

**Comment by stgraber** [comment]
Few more tweaks on the shutdown/stop logic and a rebase needed and then I think we'll be good to go!

File: lxd/api_cluster.go (hunk 29 above)
Comment (tomponline) [reply lxd/api_cluster.go:2393]: I think inst.Stop does this for us now as I recently added that.

Comment (monstermunchkin) [reply lxd/api_cluster.go:2393]: @tomponline `inst.Stop` does check if the instance is running but returns an error if it's not. I need to detect that, so I'm going with:
```go
timeout := inst.ExpandedConfig()["boot.host_shutdown_timeout"]
val, err := strconv.Atoi(timeout)
if err != nil {
	val = 30
}

err = inst.Shutdown(time.Duration(val) * time.Second)
if err != nil {
	err = inst.Stop(false)
	if err != nil && inst.IsRunning() {
		return errors.Wrapf(err, "Failed to stop instance %q", inst.Name())
	}
}
```

Comment (tomponline) [reply lxd/api_cluster.go]: @monstermunchkin to avoid unnecessary calls to the instance's status logic (which can be expensive if done repeatedly for many instances), I would be tempted to replace all occurrences of:

```go
return fmt.Errorf("The instance is already stopped")
```

With a dedicated error var that you can then check as being the cause of the error returned from both `inst.Shutdown` and `inst.Stop`, rather than calling `inst.IsRunning()` again. Effectively getting the `Stop` and `Shutdown` functions to tell you the instance is already stopped.

Comment (monstermunchkin) [reply lxd/api_cluster.go]: @tomponline that's a good idea. This should be in a seprate PR though. I'll create one.

Comment (monstermunchkin) [reply lxd/api_cluster.go]: @tomponline I've rebased the PR and added the changes.

**Comment by monstermunchkin** [comment]
jenkins: test this please
